GliaWidgets/Sources/CallVisualizer/CallVisualizer.Coordinator.swift

```diff
@@ -0,0 +1,49 @@
+import Foundation
+import UIKit
+
+extension CallVisualizer {
+    final class Coordinator {
+
+        init(
+            viewFactory: ViewFactory
+        ) {
+            self.viewFactory = viewFactory
+        }
+
+        func offerScreenShare(
+            with conf: ScreenShareOfferAlertConfiguration,
+            accepted: @escaping () -> Void,
+            declined: @escaping () -> Void
+        ) {
+            let alert = AlertViewController(
+                kind: .screenShareOffer(conf, accepted: accepted, declined: declined),
+                viewFactory: viewFactory
+            )
+            presenter.present(alert, animated: true, completion: nil)
+        }
+
+        func presentCallVisualizerViewController() {
+
+            presenter.present(
+                CallVisualizer.EngagementViewController(),
+                animated: true
+            )
+        }
+
+        // MARK: - Private
+
+        private let viewFactory: ViewFactory
+
+        private var presenter: UIViewController {
+            if var topController = UIApplication.shared.windows.first(where: { $0.isKeyWindow })?.rootViewController {
+                while let presentedViewController = topController.presentedViewController {
+                    topController = presentedViewController
+                }
+
+                return topController
```

Comment:
I see two problems here:
1. This may crash integrator's application, because we are accessing are controlled by integrator. This means if `topController` will be `UIAlertController`, presentation on it will cause runtime exception that will result in app crash.
2. UIApplication.shared is used directly. This is not testable. Instead Coordinator should have Environment inside which will be UIKitBased.UIApplication that provides access to windows.

Reply:
fixed - 4afa679cdc62918be3038a7caa4b0b3a7f611272